syne_tune/optimizer/schedulers/searchers/kde_searcher.py

```diff
@@ -0,0 +1,275 @@
+                res[k] = domain.cast(
+                    inv_numerize(
+                        values=feature_vector[curr_pos],
+                        domain=domain,
+                        categorical_map=inv_categorical_maps.get(k, {})
+                    )
+                )
+                curr_pos += 1
+            else:
+                res[k] = domain
+        return res
+
+    def to_objective(self, result: Dict):
+        if self.mode == 'min':
+            return result[self.metric_name]
+        elif self.mode == 'max':
+            return -result[self.metric_name]
+
+    def _update(self, trial_id: str, config: Dict, result: Dict):
+        self.X.append(self.to_feature(
+            config=config,
+            configspace=self.configspace,
+            categorical_maps=self.categorical_maps,
+        ))
+        self.y.append(self.to_objective(result))
+        self.train_kde(np.array(self.X), np.array(self.y))
+
+    def get_config(self, **kwargs):
+        # if not done with points_to_evaluate, pick the next one
+        if self.points_to_evaluate is not None and self.num_evaluations < len(self.points_to_evaluate):
+            suggestion = self.points_to_evaluate[self.num_evaluations]
+        else:
+            if self.good_kde is None and self.bad_kde is None:
+                # if not enough suggestion made, sample randomly
+                suggestion = {k: v.sample()
+                if isinstance(v, sp.Domain) else v for k, v in self.configspace.items()}
+            elif np.random.rand() < self.random_fraction:
+                suggestion = {k: v.sample()
+                if isinstance(v, sp.Domain) else v for k, v in self.configspace.items()}
+            else:
+                l = self.good_kde.pdf
+                g = self.bad_kde.pdf
+
+                acquisition_function = lambda x: max(1e-32, g(x)) / max(l(x), 1e-32)
+
+                current_best = None
+                val_current_best = None
+                for i in range(self.num_candidates):
+                    idx = np.random.randint(0, len(self.good_kde.data))
+                    mean = self.good_kde.data[idx]
+                    candidate = []
+
+                    for m, bw, t in zip(mean, self.good_kde.bw, self.vartypes):
+                        bw = max(bw, self.min_bandwidth)
+                        vartype = t[0]
+                        domain = t[1]
+                        if vartype == 'c':
+                            # continuous parameter
+                            bw = self.bandwidth_factor * bw
+                            candidate.append(sps.truncnorm.rvs(-m / bw, (1 - m) / bw, loc=m, scale=bw))
+                        else:
+                            # categorical or integer parameter
+                            if np.random.rand() < (1 - bw):
+                                candidate.append(m)
+                            else:
+                                if vartype == 'o':
```

Comment:
Is this again a random sample from the domain? If so, why not use domain.sample() ?

Reply:
This samples from the internal domain of the KDE model